system-test/storage.ts

```diff
@@ -1901,16 +1852,51 @@ describe('storage', () => {
           file = bucketNonAllowList.file(file.name);
         });
 
-        function doubleTest(testFunction: Function) {
+        /**
+         * A type for requester pays functions. The `options` parameter
+         *
+         * Using `typeof USER_PROJECT_OPTIONS` ensures the function should
+         * support requester pays options, otherwise will fail early at build-
+         * time rather than runtime.
+         */
+        type requesterPaysFunction<
+          T = {} | typeof USER_PROJECT_OPTIONS,
+          R = {} | void
+        > = (options: T) => Promise<R>;
+
+        /**
+         * Accepts a function and runs 2 tests - a test where the requester pays
+         * option is absent and another where it is present:
+         * - The missing requester pays test will assert the expected error
+         * - The added request pays test will return the function's result
+         *
+         * @param testFunction The function/method to test.
+         * @returns The result of the successful request pays operation.
+         */
+        async function requesterPaysDoubleTest<F extends requesterPaysFunction>(
+          testFunction: F
+        ): Promise<ReturnType<F>> {
           const failureMessage =
             'Bucket is a requester pays bucket but no user project provided.';
 
-          return (done: Function) => {
-            testFunction({}, (err: Error) => {
-              assert(err.message.indexOf(failureMessage) > -1);
-              testFunction(USER_PROJECT_OPTIONS, done);
-            });
-          };
+          let expectedError: unknown = null;
+
+          try {
+            // Should raise an error on requester pays bucket
+            await testFunction({});
+          } catch (e) {
+            expectedError = e;
+          }
+
+          assert(expectedError instanceof Error);
+          assert(
+            expectedError.message.includes(failureMessage),
```

Comment:
In light of some conversation that was had, is it currently possible to check the code here instead of the message?

Reply:
Yea, I wish... it turns out the code isn't returned properly upstream (just like the internal note on the Java client when this message broke all the things...)

Reply:
I'll file a bug internally to fix w/ the team tagged, this way we can update the tests with the code once its fixed upstream